backend/src/v2/driver/driver.go

```diff
@@ -2006,7 +2028,253 @@ func resolveUpstreamArtifacts(cfg resolveUpstreamOutputsConfig) (*pipelinespec.A
+			if len(tempSelectors) > 0 {
+				tempSubTaskName = tempSelectors[len(tempSelectors)-1].ProducerSubtask
+				tempOutputKey = tempSelectors[len(tempSelectors)-1].OutputArtifactKey
+			}
+		}
+
+	} else {
+		producerTaskValue := parentTask.GetExecution().GetCustomProperties()["parameter_producer_task"]
+		if producerTaskValue != nil {
+			tempOutputParametersMap := make(map[string]*pipelinespec.DagOutputsSpec_DagOutputParameterSpec)
+			for name, value := range producerTaskValue.GetStructValue().GetFields() {
+				outputSpec := &pipelinespec.DagOutputsSpec_DagOutputParameterSpec{}
+				err := protojson.Unmarshal([]byte(value.GetStringValue()), outputSpec)
+				if err != nil {
+					return "", "", err
+				}
+				tempOutputParametersMap[name] = outputSpec
+			}
+			glog.V(4).Infof("tempOutputParametersMap: %#v", tempOutputParametersMap)
+			switch tempOutputParametersMap[tempOutputKey].Kind.(type) {
+			case *pipelinespec.DagOutputsSpec_DagOutputParameterSpec_ValueFromParameter:
+				tempSubTaskName = tempOutputParametersMap[tempOutputKey].GetValueFromParameter().GetProducerSubtask()
+				tempOutputKey = tempOutputParametersMap[tempOutputKey].GetValueFromParameter().GetOutputParameterKey()
+			case *pipelinespec.DagOutputsSpec_DagOutputParameterSpec_ValueFromOneof:
+				// When OneOf is specified in a pipeline, the output of only 1
+				// task is consumed even though there may be more than 1 task
+				// output set. In this case we will attempt to grab the first
+				// successful task output.
+				paramSelectors := tempOutputParametersMap[tempOutputKey].GetValueFromOneof().GetParameterSelectors()
+				glog.V(4).Infof("paramSelectors: %v", paramSelectors)
+				// Since we have the tasks map, we can iterate through the
+				// parameterSelectors if the ProducerSubTask is not present in
+				// the task map and then assign the new OutputParameterKey only
+				// if it exists.
+				successfulOneOfTask := false
+				for !successfulOneOfTask {
+					for _, paramSelector := range paramSelectors {
+						tempSubTaskName = paramSelector.GetProducerSubtask()
+						// Used just for retrieval since we lookup the task in the map
+						updatedSubTaskName := metadata.UpdateTaskNameWithDagID(tempSubTaskName, parentTask.GetID())
+						glog.V(4).Infof("subTaskName with Dag ID from paramSelector: %v", updatedSubTaskName)
+						glog.V(4).Infof("outputParameterKey from paramSelector: %v", paramSelector.GetOutputParameterKey())
+						if subTask, ok := tasks[updatedSubTaskName]; ok {
+							subTaskState := subTask.GetExecution().LastKnownState.String()
+							glog.V(4).Infof("subTask: %w , subTaskState: %v", updatedSubTaskName, subTaskState)
+							if subTaskState == "CACHED" || subTaskState == "COMPLETE" {
+								tempOutputKey = paramSelector.GetOutputParameterKey()
+								successfulOneOfTask = true
+								break
+							}
+						}
+					}
+					if !successfulOneOfTask {
+						return "", "", fmt.Errorf("processing OneOf: No successful task found")
```

Comment:
If there isn't a successful task, does Argo Workflows actually get to executing the one of consumer? I'm thinking this would be hard otherwise for the user to trace down why their pipeline failed since this error is in the driver.